File: src/main/java/ar/edu/utn/frba/tacs/tp/api/herocardsgame/controllers/CardsController.java
package ar.edu.utn.frba.tacs.tp.api.herocardsgame.controllers;

import ar.edu.utn.frba.tacs.tp.api.herocardsgame.models.game.Card;
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.utils.FileConstructorUtils;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import java.util.Arrays;
import java.util.List;

@Controller
public class CardsController {
    
  @GetMapping("/users/cards")
  public ResponseEntity<List<Card>> getCards() {
    Card batman = FileConstructorUtils.createFromFile("src/test/resources/json/card/Batman.json", Card.class);
    Card flash = FileConstructorUtils.createFromFile("src/test/resources/json/card/Flash.json", Card.class);
    return ResponseEntity.status(HttpStatus.OK).body(Arrays.asList(batman, flash));
  }

  @PostMapping("/admin/cards")
  public ResponseEntity<Card> createCard(@RequestBody Card card) {
    return ResponseEntity.status(HttpStatus.CREATED).body(card);
  }

  @PutMapping("/admin/cards/{cardId}")
  public ResponseEntity updateCard(@PathVariable("cardId") String cardId, @RequestBody Card card) {
    return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
  }

  @DeleteMapping("/admin/cards/{cardId}")
  public ResponseEntity deleteCard(@PathVariable("cardId") String cardId) {
    return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
  }

  /**
   * Should present the card with the Image and attributes
   * @param cardId
   */

  @GetMapping("/users/cards")
  @RequestMapping(method = RequestMethod.HEAD, value = {"/users/cards/{cardId}"})
  public void presentCard(@PathVariable("cardId") String cardId) {
    System.out.println(cardId);
  }

}

File: src/main/kotlin/ar/edu/utn/frba/tacs/tp/api/herocardsgame/controllers/CardsController.kt
package ar.edu.utn.frba.tacs.tp.api.herocardsgame.controllers;

import ar.edu.utn.frba.tacs.tp.api.herocardsgame.exception.ElementNotFoundException
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.integration.CardIntegration
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.models.game.Card
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMethod

@Controller
@CrossOrigin(origins = ["http://localhost:5000"], allowedHeaders = ["*"])
class CardsController(val cardIntegration: CardIntegration) :
    AbstractController<CardsController>(CardsController::class.java) {

    /**
     * @param cardId
     * @return card
     */
    @GetMapping("/cards/{card-id}")
    fun getCard(@PathVariable("card-id") cardId: String): ResponseEntity<Card> =
        try {
            reportRequest(
                method = RequestMethod.GET,
                path = "/cards/{card-id}",
                body = null,
                pathVariables = hashMapOf("card-id" to cardId)
            )
            val response = cardIntegration.getCardById(cardId)
            reportResponse(HttpStatus.OK, response)
        } catch (e: ElementNotFoundException) {
            reportError(e, HttpStatus.NOT_FOUND)
        }

    /**
     * @param cardName
     * @return list of card
     */
    @GetMapping("/cards/search/{card-name}")
    fun getCardByName(@PathVariable("card-name") cardName: String): ResponseEntity<List<Card>> =
        try {
            reportRequest(
                method = RequestMethod.GET,
                path = "/cards/search/{card-name}",
                body = null,
                pathVariables = hashMapOf("card-name" to cardName)
            )
            val response = cardIntegration.searchCardByName(cardName)
            reportResponse(HttpStatus.OK, response)
        } catch (e: ElementNotFoundException) {
            reportError(e, HttpStatus.NOT_FOUND)
        }

    /**
     * @return list of saved cards
     */
    @GetMapping("/cards")
    fun getSavedCards(): ResponseEntity<List<Card>> {
        reportRequest(
            method = RequestMethod.GET,
            path = "/cards",
            body = null
        )
        val response = cardIntegration.getSavedCards()
        return reportResponse(HttpStatus.OK, response)
    }

}